kotlin/src/main/java/feign/kotlin/CoroutineFeign.java

```diff
@@ -181,178 +149,51 @@ public <T> T target(Target<T> target, C context) {
-        return result;
-      };
-    }
-
-    // from SynchronousMethodHandler
-    long elapsedTime(long start) {
-      return TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
-    }
-
-    private Decoder stageDecode(
-                                ThreadLocal<AsyncInvocation<C>> activeContext,
-                                Logger logger,
-                                Level logLevel,
-                                AsyncResponseHandler responseHandler) {
-      return (response, type) -> {
-        final AsyncInvocation<C> invocationContext = activeContext.get();
-
-        final CompletableFuture<Object> result = new CompletableFuture<>();
-
-        invocationContext
-            .responseFuture()
-            .whenComplete(
-                (r, t) -> {
-                  final long elapsedTime = elapsedTime(invocationContext.startNanos());
-
-                  if (t != null) {
-                    if (logLevel != Logger.Level.NONE && t instanceof IOException) {
-                      final IOException e = (IOException) t;
-                      logger.logIOException(
-                          invocationContext.configKey(), logLevel, e, elapsedTime);
-                    }
-                    result.completeExceptionally(t);
-                  } else {
-                    responseHandler.handleResponse(
-                        result,
-                        invocationContext.configKey(),
-                        r,
-                        invocationContext.underlyingType(),
-                        elapsedTime);
-                  }
-                });
-
-        result.whenComplete(
-            (r, t) -> {
-              if (result.isCancelled()) {
-                invocationContext.responseFuture().cancel(true);
-              }
-            });
-
-        if (invocationContext.isAsyncReturnType()) {
-          return result;
-        }
-        try {
-          return result.join();
-        } catch (final CompletionException e) {
-          final Response r = invocationContext.responseFuture().join();
-          Throwable cause = e.getCause();
-          if (cause == null) {
-            cause = e;
-          }
-          throw new AsyncJoinException(r.status(), cause.getMessage(), r.request(), cause);
-        }
-      };
+      AsyncFeign<C> asyncFeign = (AsyncFeign<C>) AsyncFeign.asyncBuilder()
+          .logLevel(logLevel)
+          .client((AsyncClient<Object>) client)
+          .decoder(decoder)
+          .contract(contract)
+          .retryer(retryer)
+          .logger(logger)
+          .encoder(encoder)
+          .queryMapEncoder(queryMapEncoder)
+          .options(options)
+          .requestInterceptors(requestInterceptors)
+          .responseInterceptor(responseInterceptor)
+          .invocationHandlerFactory(invocationHandlerFactory)
+          .defaultContextSupplier((AsyncContextSupplier<Object>) defaultContextSupplier)
+          .methodInfoResolver(KotlinMethodInfo::createInstance)
```

Comment:
shouldn't this be extracted to an attribute on the builder and let people override if needed?

Also, that is necessary to allow coroutine `methodInfoResolver` to be enriched

Reply:
Do you mean to allow the CoroutineBuilder to set the methodInfoResolver, and also provide an option to enrich the methodInfoResolver using Capability?

Reply:
I created a new PR #1762

Reply:
Yes! Perfect